### PR TITLE
ARROW-16721: [C++] Drop support for bundled Thrift < 0.13

### PR DESCRIPTION
**Pull request by kou**
Our bundled Thrift is 0.16.0. Users can use other version but will not
use 0.13 or earlier.

This change also cleans FindThrift.cmake up.

This change also fixes build errors with -DARROW_HIVESERVER2=ON. We
may be able to remove cpp/src/arrow/dbi/hiveserver2/. It seems that
nobody uses it.

### CODE REVIEW

File: cpp/src/arrow/dbi/hiveserver2/CMakeLists.txt [hunk 1]
```diff
@@ -31,29 +31,32 @@ set(ARROW_HIVESERVER2_SRCS
     types.cc
     util.cc)
 
+set(HIVESERVER2_THRIFT_SRC_DIR "${ARROW_BINARY_DIR}/src/arrow/dbi/hiveserver2")
```
Comment (pitrou) [inline cpp/src/arrow/dbi/hiveserver2/CMakeLists.txt:34]: I think our hiveserver adapter doesn't even build currently, do we want to bother fixing it?

**Comment by pitrou** [comment]
There's some compatibility code lying around (look for PARQUET_THRIFT_USE_BOOST and FORCE_BOOST_SMART_PTR), we should probably remove it as well.

**Comment by pitrou** [comment]
> We
> may be able to remove cpp/src/arrow/dbi/hiveserver2/. It seems that
> nobody uses it.

+1

**Comment by kou** [comment]
I've started a discussion for removing `cpp/src/arrow/dbi/hiveserver2/`:
[C++] Can we remove cpp/src/arrow/dbi/hiveserver2? https://lists.apache.org/thread/70qv1q9krx7ztk35tzxq8jp11vq5b5zt

**Comment by kou** [comment]
> There's some compatibility code lying around (look for PARQUET_THRIFT_USE_BOOST and FORCE_BOOST_SMART_PTR), we should probably remove it as well.

Good catch!
I've removed them because they are for Thrift < 0.11 that isn't supported.

**Comment by kou** [comment]
> I've started a discussion for removing `cpp/src/arrow/dbi/hiveserver2/`:
> [C++] Can we remove cpp/src/arrow/dbi/hiveserver2? https://lists.apache.org/thread/70qv1q9krx7ztk35tzxq8jp11vq5b5zt

I'll remove `cpp/src/arrow/dbi/hiveserver2/` in a follow-up pull request once we get a consensus of it.

**Comment by pitrou** [comment]
The R CI failures look related.

**Comment by kou** [comment]
I've fixed R failures but new R failures are happen. The new failures are also happen on master: https://github.com/apache/arrow/runs/6747296145?check_suite_focus=true

It seems that duckdb is updated to 0.3.4 from 0.3.2.
Cc: @nealrichardson

**Comment by nealrichardson** [comment]
> I've fixed R failures but new R failures are happen. The new failures are also happen on master: https://github.com/apache/arrow/runs/6747296145?check_suite_focus=true
> 
> It seems that duckdb is updated to 0.3.4 from 0.3.2. Cc: @nealrichardson

Yeah I noticed that too. We'll figure out something today.

**Comment by nealrichardson** [comment]
https://github.com/apache/arrow/pull/13323 is merged; rebase should make this go away

**Comment by kou** [comment]
Thanks!

This is ready to merge.

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 2]
```diff
@@ -1011,7 +983,7 @@ if(ARROW_BUILD_INTEGRATION
   set(ARROW_USE_BOOST TRUE)
   set(ARROW_BOOST_REQUIRE_LIBRARY TRUE)
 elseif(ARROW_GANDIVA
-       OR (ARROW_WITH_THRIFT AND THRIFT_REQUIRES_BOOST)
+       OR ARROW_WITH_THRIFT
```
Comment (pitrou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake:986]: So boost is always required by Thrift nevertheless?

Comment (kou) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake:986]: Yes. The condition was wrong. We needs Boost to use Thrift.

We need Thrift only when we use Parquet or `arrow/dbi/hiveserver2/`.
Parquet uses `thrift/transport/TBufferTransports.h`: https://github.com/apache/arrow/blob/master/cpp/src/parquet/thrift_internal.h#L43
`thrift/transport/TBufferTransports.h` uses Boost: https://github.com/apache/thrift/blob/master/lib/cpp/src/thrift/transport/TBufferTransports.h#L26

**Review by pitrou — APPROVED** [review]

**Pull request merged**